File: runtime/java/src/main/java/dev/argon/verilization/java_runtime/String.java
package dev.argon.verilization.java_runtime;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.math.BigInteger;

public final class String {
    private String() {}

    public static java.lang.String fromString(java.lang.String s) {
        return s;
    }

    public static final Codec<java.lang.String> codec = new Codec<java.lang.String>() {
        @Override
        public java.lang.String read(FormatReader reader) throws IOException {
            BigInteger length = Nat.codec.read(reader);
            byte[] data = reader.readBytes(length.intValueExact());
            return new java.lang.String(data, StandardCharsets.UTF_8);
        }

        @Override
        public void write(FormatWriter writer, java.lang.String value) throws IOException {
            byte[] data = value.getBytes(StandardCharsets.UTF_8);
            Nat.codec.write(writer, BigInteger.valueOf(data.length));
            writer.writeBytes(data);
        }
    };
}

File: runtime/scala/src/main/scala/dev/argon/verilization/scala_runtime/String.scala
package dev.argon.verilization.scala_runtime

import zio.{ZIO, Chunk}
import scala.Predef.{String => SString}
import java.nio.charset.StandardCharsets

object String {
    def fromString(s: SString): String = s

    val codec: Codec[SString] = new Codec[SString] {
        override def read[R, E](reader: FormatReader[R, E]): ZIO[R, E, SString] =
            Nat.codec.read(reader).flatMap { length =>
                reader.readBytes(length.bigInteger.intValueExact)
            }.map { data => new SString(data.toArray, StandardCharsets.UTF_8) }

        override def write[R, E](writer: FormatWriter[R, E], value: SString): ZIO[R, E, Unit] = {
            val data = value.getBytes(StandardCharsets.UTF_8)
            Nat.codec.write(writer, value.length) *> writer.writeBytes(Chunk.fromArray(data))
        }
            
    }
}
